File: ios/PanoramaView.m
//
//  PanoramaView.m
//
//  Created by Rody Molenaar on 08/02/19.
//  Copyright © 2019 Lightbase B.V. All rights reserved.
//

#import <React/RCTImageLoader.h>
#import "PanoramaView.h"

@implementation PanoramaView {
    UIImage *_image;
    CTPanoramaView *_panoView;
    __weak RCTBridge *_bridge;

    RCTImageLoaderCancellationBlock _cancel;
}

- (dispatch_queue_t)methodQueue
{
    return dispatch_get_main_queue();
}

- (instancetype)initWithBridge:(RCTBridge *)bridge
{
    if ((self = [super init])) {
        _bridge = bridge;
        _cancel = nil;
    }

    return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];

    _panoView = [[CTPanoramaView alloc] initWithFrame:frame];
    _panoView.startAngle = 3.45;

    [self addSubview:_panoView];

    return self;
}

- (void)layoutSubviews
{
    float rootViewWidth = self.frame.size.width;
    float rootViewHeight = self.frame.size.height;

    _panoView.frame = CGRectMake(0, 0, rootViewWidth, rootViewHeight);
}


-(void)setImageUrl:(NSString *)imageUrl
{
    NSLog(@"%@", [NSString stringWithFormat:@"[PanoramaView] Image url: %@", imageUrl]);

    __weak PanoramaView *weakSelf = self;

    if (_cancel != nil) {
        _cancel();
        _cancel = nil;
    }

    if (imageUrl.length && _bridge) {
        NSLog(@"[PanoramaView] Getting ready to load.");

        _cancel = [[_bridge moduleForName:@"ImageLoader" lazilyLoadIfNecessary:YES] loadImageWithURLRequest:[RCTConvert NSURLRequest:imageUrl] callback:^(NSError *error, UIImage *image) {

            if (error) {
                NSLog(@"[PanoramaView] Loading image error: %@.", error);
                if(self -> _cancel != nil){
                    [self imageLoadingFailed];
                }
            }
            else{
                NSLog(@"[PanoramaView] Image downloaded.");

                dispatch_async([weakSelf methodQueue], ^{
                    [self imageDownloaded];
                });
                dispatch_async([weakSelf methodQueue], ^{
                    if (image) {
                        self->_panoView.image = image;
                        [self imageLoaded];
                    } else {
                        [self imageLoadingFailed];
                    }
                });
            }
            self->_cancel = nil;
        }];

    } else {
        if (_bridge == nil) {
            NSLog(@"[PanoramaView] Bridge not available.");
        }
        if (!imageUrl.length) {
            NSLog(@"[PanoramaView] Image argument not sufficient.");
        }
        if (!_bridge.imageLoader) {
            NSLog(@"[PanoramaView] Bridge image loader not available.");
        }
        NSLog(@"[PanoramaView] Image argument not sufficient or bridge image loader not available.");

        [self imageLoadingFailed];
    }
}

-(void)setEnableTouchTracking:(BOOL)enableTouchTracking
{
    if (enableTouchTracking) {
        //_panoView.controlMethod = CTPanoramaControlMethodTouch;
        _panoView.controlMethod = CTPanoramaControlMethodBoth;
    } else {
        _panoView.controlMethod = CTPanoramaControlMethodMotion;
    }
}

- (void)imageLoadingFailed {
    if (_onImageLoadingFailed) {
        _onImageLoadingFailed(nil);
    }
}

- (void)imageDownloaded {
    if (_onImageDownloaded) {
        _onImageDownloaded(nil);
    }
}

- (void)imageLoaded {
    if (_onImageLoaded) {
        _onImageLoaded(nil);
    }
}

- (void)dealloc
{
    if(_cancel != nil){
        _cancel();
        _cancel = nil;
    }
}

- (void)willMoveToSuperview:(nullable UIView *)newSuperview;
{
    // cleanup on deallocate
    // cancel any request in progress
    if(newSuperview == nil){
        if(_cancel != nil){
            _cancel();
            _cancel = nil;
       }
    }

    [super willMoveToSuperview:newSuperview];
}

@end

File: ios/PanoramaView.swift
import Foundation


@objc public class PanoramaView: UIView {


    // MARK: Public properties

    @objc public var bridge : RCTBridge? = nil
    @objc public var onImageLoadingFailed: RCTDirectEventBlock? = nil
    @objc public var onImageDownloaded: RCTDirectEventBlock? = nil
    @objc public var onImageLoaded: RCTDirectEventBlock? = nil

    @objc public var enableTouchTracking: Bool = true {
        didSet {
            if(enableTouchTracking){
                self.panoramaView?.controlMethod = .both;
            }
            else{
                self.panoramaView?.controlMethod = .motion;
            }
        }
    }
    
    private static var metalSize: Int = {
        var maxTexSize = 8192
        if let device = MTLCreateSystemDefaultDevice() {
            if #available(iOS 11.0, *) {
                if device.supportsFeatureSet(.iOS_GPUFamily4_v1) || device.supportsFeatureSet(.iOS_GPUFamily3_v1) {
                    maxTexSize = 16384
                } else if device.supportsFeatureSet(.iOS_GPUFamily2_v2) || device.supportsFeatureSet(.iOS_GPUFamily1_v2) {
                    maxTexSize = 8192
                } else {
                    maxTexSize = 4096
                }
            } else {
                if device.supportsFeatureSet(.iOS_GPUFamily2_v2) || device.supportsFeatureSet(.iOS_GPUFamily1_v2) {
                    maxTexSize = 8192
                } else {
                    maxTexSize = 4096
                }
            }
        }
        return maxTexSize
    }()

    @objc public var imageUrl: String? = nil {
        didSet {
            if(!(imageUrl?.isEmpty ?? true)){

                // this runs in the UI thread so it is fine to call it here
                if(self.bridge == nil){
                    self.onImageLoadingFailed?(["error": "Bridge is not ready or not set."])
                    return
                }

                // ugly part of using RN's objective C code in Swift
                let loader = self.bridge?.module(forName: "ImageLoader", lazilyLoadIfNecessary: true) as! RCTImageLoader

                self.cancel?()
                self.cancel = nil

                guard let request = RCTConvert.nsurlRequest(imageUrl) else {
                    return
                }

                self.cancel = loader.loadImage(with: request, callback: { (error, image) in

                    DispatchQueue.main.async {

                        // do nothing if we have cancelled the request
                        if(self.cancel == nil){
                            return
                        }

                        if (error != nil) {
                            self.onImageLoadingFailed?(["error": error!.localizedDescription])
                        }
                        else{
                            self.onImageDownloaded?(nil)

                            if let _image = image {
                                var resized = _image
                                let size = _image.size
                                
                                // Resize image if too big or the viewer will choke
                                // Get max metal supported size to prevent
                                // 'Texture Descriptor Validation MTLTextureDescriptor has width (XXXX) greater than the maximum allowed size of 16384.' errors
                                let targetSize = CGSize(width: PanoramaView.metalSize, height: PanoramaView.metalSize)
                                
                                if size.width > targetSize.width || size.height > targetSize.height {
                                    
                                    let widthRatio  = targetSize.width / size.width
                                    let heightRatio = targetSize.height / size.height

                                    // Figure out what our orientation is, and use that to form the rectangle
                                    var newSize: CGSize
                                    if(widthRatio > heightRatio) {
                                       newSize = CGSize(width: size.width * heightRatio, height: size.height * heightRatio)
                                    } else {
                                       newSize = CGSize(width: size.width * widthRatio, height: size.height * widthRatio)
                                    }
                                    
                                    
                                    // This is the rect that we've calculated out and this is what is actually used below
                                    let rect = CGRect(origin: .zero, size: newSize)

                                    // Actually do the resizing to the rect using the ImageContext stuff
                                    UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
                                    
                                    _image.draw(in: rect)
                                    
                                    resized = UIGraphicsGetImageFromCurrentImageContext() ?? resized
                                    
                                    UIGraphicsEndImageContext()
                                }
                                
                                self.panoramaView?.image = resized
                                self.onImageLoaded?(nil)
                            }
                            else{
                                self.onImageLoadingFailed?(["error": "Image was empty or failed to load."])
                            }
                        }

                        self.cancel = nil
                    }
                })
            }
            else{
                self.cancel?()
                self.cancel = nil
            }
        }
    }



    // MARK: Private properties

    private var panoramaView : CTPanoramaView? = nil
    private var cancel : RCTImageLoaderCancellationBlock? = nil


    // MARK: Class lifecycle methods

    public override init(frame: CGRect) {
        super.init(frame: frame)
        let view = CTPanoramaView(frame: frame)
        view.startAngle = 3.45
        view.controlMethod = .both;

        self.panoramaView = view
        self.addSubview(view)
    }


    public required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    deinit {
        self.cancel?()
        self.cancel = nil
    }

    public override func layoutSubviews() {
        super.layoutSubviews()
        self.panoramaView?.frame = CGRect(x: 0, y: 0, width: frame.size.width, height: frame.size.height)
        self.panoramaView?.setNeedsDisplay()
    }

    public override func willMove(toSuperview newSuperview: UIView?){
        super.willMove(toSuperview: newSuperview)

        // cleanup/cancel downloads on view removal
        if(newSuperview == nil){
            self.cancel?()
            self.cancel = nil
        }
    }
}
